src/nvim/undo.c

```diff
@@ -3029,15 +3029,14 @@ u_header_T *u_force_get_undo_header(buf_T *buf)
     curbuf = buf;
     // Args are tricky: this means replace empty range by empty range..
     u_savecommon(0, 1, 1, true);
-    curbuf = save_curbuf;
-
     uhp = buf->b_u_curhead;
     if (!uhp) {
       uhp = buf->b_u_newhead;
       if (get_undolevel() > 0 && !uhp) {
         abort();
       }
     }
+    curbuf = save_curbuf;
```

Comment:
I did some testing and everything seems to work, even without this change. Is this change needed anymore ?

Reply:
might be worth investigating how easy it would be to pass a buffer to u_savecommon

Reply:
Yep, definitely, adding that to my `decurbuf` roadmap then.

Reply:
In the absence of this change, the following steps can cause a crash similar to #12873.

```vim
nvim -u NORC
:vsplit file
:let bufnr = 1
:call nvim_buf_set_option(bufnr, "undofile", v:false)
:call nvim_buf_set_option(bufnr, "undolevels", -1)
:let g:mark_ns = nvim_create_namespace('myns')
:let g:mark_id = nvim_buf_set_extmark(1, g:mark_ns, 0, 0, {})
```

Reply:
Since `uhp` cannot be `NULL` except when `undoleves` is -1, shouldn't the interior of the following `if` clause never be executed? Doesn't the internal execution of the `if` clause below imply that the creation or deletion of the `extmark`, without editing, creates an entry for undo/redo?

https://github.com/neovim/neovim/blob/4f8d98e583beb4c1abd5d57b9898548396633030/src/nvim/undo.c#L3024-L3041